// @balmy/sdk/dist/services/quotes/quote-sources/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checksum = exports.calculateAllowanceTarget = exports.addQuoteSlippage = exports.failed = void 0;
const viem_1 = require("viem");
const utils_1 = require("../../../shared/utils");
const errors_1 = require("../errors");
const constants_1 = require("../../../shared/constants");
function failed(metadata, chain, sellToken, buyToken, error) {
    const chainId = typeof chain === 'number' ? chain : chain.chainId;
    throw new errors_1.FailedToGenerateQuoteError(metadata.name, chainId, sellToken, buyToken, error);
}
exports.failed = failed;
function addQuoteSlippage(quote, type, slippagePercentage) {
    return type === 'sell'
        ? {
            ...quote,
            type,
            minBuyAmount: (0, utils_1.subtractPercentage)(quote.buyAmount, slippagePercentage, 'up'),
            maxSellAmount: quote.sellAmount,
        }
        : {
            ...quote,
            type,
            maxSellAmount: (0, utils_1.addPercentage)(quote.sellAmount, slippagePercentage, 'up'),
            minBuyAmount: quote.buyAmount,
        };
}
exports.addQuoteSlippage = addQuoteSlippage;
function calculateAllowanceTarget(sellToken, allowanceTarget) {
    return (0, utils_1.isSameAddress)(sellToken, constants_1.Addresses.NATIVE_TOKEN) ? constants_1.Addresses.ZERO_ADDRESS : allowanceTarget;
}
exports.calculateAllowanceTarget = calculateAllowanceTarget;
function checksum(address) {
    return (0, viem_1.getAddress)(address);
}
exports.checksum = checksum;
