// @balmy/sdk/dist/services/logs/loggers/console-logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleLogger = void 0;
const utils_1 = require("../utils");
class ConsoleLogger {
    constructor(name, level) {
        this.name = name;
        this.level = level;
    }
    log(message, ...optionalParams) {
        this.print('LOG', message, optionalParams);
    }
    debug(message, ...optionalParams) {
        this.print('DEBUG', message, optionalParams);
    }
    info(message, ...optionalParams) {
        this.print('INFO', message, optionalParams);
    }
    warn(message, ...optionalParams) {
        this.print('WARN', message, optionalParams);
    }
    error(message, ...optionalParams) {
        this.print('ERROR', message, optionalParams);
    }
    print(level, message, optionalParams) {
        if ((0, utils_1.shouldPrintLog)(level, this.level)) {
            if (optionalParams.length > 1) {
                return console.log(new Date().toISOString() + ' [' + level + '] (' + this.name + '): ' + (message ?? ''), ...optionalParams);
            }
            else {
                return console.log(new Date().toISOString() + ' [' + level + '] (' + this.name + '): ' + (message ?? ''));
            }
        }
    }
}
exports.ConsoleLogger = ConsoleLogger;
