// @balmy/sdk/dist/services/providers/provider-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProviderService = void 0;
const viem_1 = require("viem");
const utils_1 = require("./utils");
class ProviderService {
    constructor(source) {
        this.source = source;
        // Viem clients have a lot of state and they even do some polling at regular intervals
        // That's why we'll only create one client per chain, and then re-use it
        this.viemPublicClients = new Map();
    }
    supportedChains() {
        return this.source.supportedChains();
    }
    getViemPublicClient({ chainId }) {
        if (!this.viemPublicClients.has(chainId)) {
            const transport = this.getViemTransport({ chainId });
            const chain = (0, utils_1.getViemChain)(chainId);
            const client = (0, viem_1.createPublicClient)({ chain, transport });
            this.viemPublicClients.set(chainId, client);
        }
        return this.viemPublicClients.get(chainId);
    }
    getViemTransport({ chainId }) {
        return this.source.getViemTransport({ chainId });
    }
}
exports.ProviderService = ProviderService;
