// @balmy/sdk/dist/services/providers/provider-sources/blast-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.blastSupportedChains = exports.buildBlastRPCUrl = exports.BlastProviderSource = void 0;
const _chains_1 = require("../../../chains");
const base_http_provider_1 = require("./base/base-http-provider");
const PLACEHOLDER = '{{ PLACEHOLDER }}';
const SUPPORTED_CHAINS = {
    [_chains_1.Chains.ETHEREUM.chainId]: `https://eth-mainnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.ETHEREUM_SEPOLIA.chainId]: `https://eth-sepolia.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.BNB_CHAIN.chainId]: `https://bsc-mainnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.POLYGON.chainId]: `https://polygon-mainnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.POLYGON_MUMBAI.chainId]: `https://polygon-testnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.ARBITRUM.chainId]: `https://arbitrum-one.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.ASTAR.chainId]: `https://astar.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.OPTIMISM.chainId]: `https://optimism-mainnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.LINEA.chainId]: `https://linea-mainnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.BASE.chainId]: `https://base-mainnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.FANTOM.chainId]: `https://fantom-mainnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.AVALANCHE.chainId]: `https://ava-mainnet.blastapi.io${PLACEHOLDER}/ext/bc/C/rpc`,
    [_chains_1.Chains.GNOSIS.chainId]: `https://gnosis-mainnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.POLYGON_ZKEVM.chainId]: `https://polygon-zkevm-mainnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.MOONBEAM.chainId]: `https://moonbeam.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.MOONRIVER.chainId]: `https://moonriver.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.OKC.chainId]: `https://oktc-mainnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.MODE.chainId]: `https://mode-mainnet.blastapi.io${PLACEHOLDER}`,
    [_chains_1.Chains.SCROLL.chainId]: `https://scroll-mainnet.public.blastapi.io${PLACEHOLDER}`,
};
class BlastProviderSource extends base_http_provider_1.BaseHttpProvider {
    constructor(key, onChains) {
        super();
        this.key = key;
        this.supported = onChains ?? blastSupportedChains();
    }
    supportedChains() {
        return this.supported;
    }
    calculateUrl(chainId) {
        return buildBlastRPCUrl({ chainId, apiKey: this.key });
    }
}
exports.BlastProviderSource = BlastProviderSource;
function buildBlastRPCUrl({ chainId, apiKey }) {
    const url = SUPPORTED_CHAINS[chainId];
    const toReplace = apiKey ? `/${apiKey}` : '';
    return url.replace(PLACEHOLDER, toReplace);
}
exports.buildBlastRPCUrl = buildBlastRPCUrl;
function blastSupportedChains() {
    return Object.keys(SUPPORTED_CHAINS).map(Number);
}
exports.blastSupportedChains = blastSupportedChains;
