// @balmy/sdk/dist/services/providers/provider-sources/infura-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.infuraSupportedChains = exports.buildInfuraRPCUrl = exports.InfuraProviderSource = void 0;
const _chains_1 = require("../../../chains");
const base_http_provider_1 = require("./base/base-http-provider");
const SUPPORTED_CHAINS = {
    [_chains_1.Chains.ETHEREUM.chainId]: 'https://mainnet.infura.io/v3/',
    [_chains_1.Chains.ETHEREUM_GOERLI.chainId]: 'https://goerli.infura.io/v3/',
    [_chains_1.Chains.ETHEREUM_SEPOLIA.chainId]: 'https://sepolia.infura.io/v3/',
    [_chains_1.Chains.ARBITRUM.chainId]: 'https://arbitrum-mainnet.infura.io/v3/',
    [_chains_1.Chains.AURORA.chainId]: 'https://aurora-mainnet.infura.io/v3/',
    [_chains_1.Chains.AVALANCHE.chainId]: 'https://avalanche-mainnet.infura.io/v3/',
    [_chains_1.Chains.CELO.chainId]: 'https://celo-mainnet.infura.io/v3/',
    [_chains_1.Chains.OPTIMISM.chainId]: 'https://optimism-mainnet.infura.io/v3/',
    [_chains_1.Chains.POLYGON.chainId]: 'https://polygon-mainnet.infura.io/v3/',
    [_chains_1.Chains.POLYGON_MUMBAI.chainId]: 'https://polygon-mumbai.infura.io/v3/',
};
class InfuraProviderSource extends base_http_provider_1.BaseHttpProvider {
    constructor(key, onChains) {
        super();
        this.key = key;
        this.supported = onChains ?? infuraSupportedChains();
    }
    supportedChains() {
        return this.supported;
    }
    calculateUrl(chainId) {
        return buildInfuraRPCUrl({ chainId, apiKey: this.key });
    }
}
exports.InfuraProviderSource = InfuraProviderSource;
function buildInfuraRPCUrl({ chainId, apiKey }) {
    return SUPPORTED_CHAINS[chainId] + apiKey;
}
exports.buildInfuraRPCUrl = buildInfuraRPCUrl;
function infuraSupportedChains() {
    return Object.keys(SUPPORTED_CHAINS).map(Number);
}
exports.infuraSupportedChains = infuraSupportedChains;
