// @balmy/sdk/dist/services/providers/provider-sources/node-real-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeRealSupportedChains = exports.buildNodeRealRPCUrl = exports.NodeRealProviderSource = void 0;
const _chains_1 = require("../../../chains");
const base_http_provider_1 = require("./base/base-http-provider");
const SUPPORTED_CHAINS = {
    [_chains_1.Chains.ETHEREUM.chainId]: 'https://eth-mainnet.nodereal.io/v1/',
    [_chains_1.Chains.ETHEREUM_GOERLI.chainId]: 'https://eth-goerli.nodereal.io/v1/',
    [_chains_1.Chains.BNB_CHAIN.chainId]: 'https://bsc-mainnet.nodereal.io/v1/',
    [_chains_1.Chains.POLYGON.chainId]: 'https://polygon-mainnet.nodereal.io/v1/',
    [_chains_1.Chains.OPTIMISM.chainId]: 'https://opt-mainnet.nodereal.io/v1/',
};
class NodeRealProviderSource extends base_http_provider_1.BaseHttpProvider {
    constructor(key, onChains) {
        super();
        this.key = key;
        this.supported = onChains ?? nodeRealSupportedChains();
    }
    supportedChains() {
        return this.supported;
    }
    calculateUrl(chainId) {
        return buildNodeRealRPCUrl({ chainId, apiKey: this.key });
    }
}
exports.NodeRealProviderSource = NodeRealProviderSource;
function buildNodeRealRPCUrl({ chainId, apiKey }) {
    return SUPPORTED_CHAINS[chainId] + apiKey;
}
exports.buildNodeRealRPCUrl = buildNodeRealRPCUrl;
function nodeRealSupportedChains() {
    return Object.keys(SUPPORTED_CHAINS).map(Number);
}
exports.nodeRealSupportedChains = nodeRealSupportedChains;
