// @balmy/sdk/dist/services/providers/provider-sources/tenderly-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tenderlySupportedChains = exports.buildTenderlyRPCUrl = exports.TenderlyProviderSource = void 0;
const _chains_1 = require("../../../chains");
const base_http_provider_1 = require("./base/base-http-provider");
const SUPPORTED_CHAINS = {
    [_chains_1.Chains.ETHEREUM.chainId]: 'https://mainnet.gateway.tenderly.co',
    [_chains_1.Chains.ETHEREUM_SEPOLIA.chainId]: 'https://sepolia.gateway.tenderly.co',
    [_chains_1.Chains.POLYGON.chainId]: 'https://polygon.gateway.tenderly.co',
    [_chains_1.Chains.POLYGON_MUMBAI.chainId]: 'https://polygon-mumbai.gateway.tenderly.co',
    [_chains_1.Chains.OPTIMISM.chainId]: 'https://optimism.gateway.tenderly.co',
    [_chains_1.Chains.BASE.chainId]: 'https://base.gateway.tenderly.co',
    [_chains_1.Chains.ARBITRUM.chainId]: 'https://arbitrum.gateway.tenderly.co',
    [_chains_1.Chains.BOBA.chainId]: 'https://boba-ethereum.gateway.tenderly.co',
    [_chains_1.Chains.BLAST.chainId]: 'https://blast.gateway.tenderly.co',
    [_chains_1.Chains.MODE.chainId]: 'https://mode.gateway.tenderly.co',
};
class TenderlyProviderSource extends base_http_provider_1.BaseHttpProvider {
    constructor(key, onChains) {
        super();
        this.key = key;
        this.supported = onChains ?? tenderlySupportedChains();
    }
    supportedChains() {
        return this.supported;
    }
    calculateUrl(chainId) {
        return buildTenderlyRPCUrl({ chainId, apiKey: this.key });
    }
}
exports.TenderlyProviderSource = TenderlyProviderSource;
function buildTenderlyRPCUrl({ chainId, apiKey }) {
    let url = SUPPORTED_CHAINS[chainId];
    if (apiKey) {
        url += `/${apiKey}`;
    }
    return url;
}
exports.buildTenderlyRPCUrl = buildTenderlyRPCUrl;
function tenderlySupportedChains() {
    return Object.keys(SUPPORTED_CHAINS).map(Number);
}
exports.tenderlySupportedChains = tenderlySupportedChains;
