// @balmy/sdk/dist/services/allowances/allowance-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllowanceService = void 0;
const timeouts_1 = require("../../shared/timeouts");
class AllowanceService {
    constructor(source) {
        this.source = source;
    }
    supportedChains() {
        return this.source.supportedChains();
    }
    async getAllowanceInChain({ chainId, token, owner, spender, config, }) {
        const result = await this.getAllowancesInChain({
            chainId,
            allowances: [{ token, owner, spender }],
            config,
        });
        return result[token][owner][spender];
    }
    async getAllowancesInChain({ chainId, allowances, config, }) {
        const result = await this.getAllowances({
            allowances: allowances.map((allowance) => ({ chainId, ...allowance })),
            config,
        });
        return result[chainId] ?? {};
    }
    async getAllowances({ allowances, config }) {
        return (0, timeouts_1.timeoutPromise)(this.source.getAllowances({
            allowances,
            config,
        }), config?.timeout);
    }
}
exports.AllowanceService = AllowanceService;
