// @balmy/sdk/dist/services/dca/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DCASwapInterval = exports.DCAPermission = exports.ActionTypeAction = void 0;
var ActionTypeAction;
(function (ActionTypeAction) {
    ActionTypeAction["CREATED"] = "created";
    ActionTypeAction["MODIFIED"] = "modified";
    ActionTypeAction["WITHDRAWN"] = "withdrawn";
    ActionTypeAction["TERMINATED"] = "terminated";
    ActionTypeAction["TRANSFERRED"] = "transferred";
    ActionTypeAction["MODIFIED_PERMISSIONS"] = "modified permissions";
    ActionTypeAction["SWAPPED"] = "swapped";
})(ActionTypeAction || (exports.ActionTypeAction = ActionTypeAction = {}));
var DCAPermission;
(function (DCAPermission) {
    DCAPermission["INCREASE"] = "INCREASE";
    DCAPermission["REDUCE"] = "REDUCE";
    DCAPermission["WITHDRAW"] = "WITHDRAW";
    DCAPermission["TERMINATE"] = "TERMINATE";
})(DCAPermission || (exports.DCAPermission = DCAPermission = {}));
var DCASwapInterval;
(function (DCASwapInterval) {
    DCASwapInterval[DCASwapInterval["ONE_MINUTE"] = 60] = "ONE_MINUTE";
    DCASwapInterval[DCASwapInterval["FIVE_MINUTES"] = 300] = "FIVE_MINUTES";
    DCASwapInterval[DCASwapInterval["FIFTEEN_MINUTES"] = 900] = "FIFTEEN_MINUTES";
    DCASwapInterval[DCASwapInterval["THIRTY_MINUTES"] = 1800] = "THIRTY_MINUTES";
    DCASwapInterval[DCASwapInterval["ONE_HOUR"] = 3600] = "ONE_HOUR";
    DCASwapInterval[DCASwapInterval["FOUR_HOURS"] = 14400] = "FOUR_HOURS";
    DCASwapInterval[DCASwapInterval["ONE_DAY"] = 86400] = "ONE_DAY";
    DCASwapInterval[DCASwapInterval["ONE_WEEK"] = 604800] = "ONE_WEEK";
})(DCASwapInterval || (exports.DCASwapInterval = DCASwapInterval = {}));
